strawberryfields/backends/fockbackend/circuit.py

```diff
@@ -135,6 +113,249 @@ def reset(self, pure=None, cutoff_dim=None, num_subsystems=None):
         else:
             self._state = ops.vacuumStateMixed(self._num_modes, self._trunc)
 
+    def apply_gate_BLAS(self, mat, modes, **kwargs):
+        """Applies a specified gate to the state.
+
+        Gate application based on custom indexing and matrix multiplication.
+        Assumes the input matrix has shape (out1, in1, ...).
+
+        This implementation uses indexing and BLAS. As per stack overflow,
+        einsum doesn't actually use BLAS but rather a c implementation. In theory
+        if reshaping is efficient this should be faster.
+
+        Args:
+            mat (array[complex]): The numeric operator to be applied to the state, of shape `[trunc]*(2*n)`
+            modes (list[int]): The list of modes to which the operator is applied on
+
+        Keyword args:
+            state (array[complex]): The state that the operator is applied to
+            pure (bool): Whether the state is pure or mixed
+            n (int): The total number of modes
+
+        Returns:
+            array[complex]: The state after application of the two-mode operation
+        """
+        # pylint: disable=too-many-branches
+
+        state = kwargs.get("state", self._state)
+        pure = kwargs.get("pure", self._pure)
+        n = kwargs.get("n", self._num_modes)
```

Comment:
Currently, I'm still grasping how the internals of SF work, what would be the use cases when these would be passed as a keyword argument (and not default to the values of `self._state`, `self._pure` and `self._num_modes`)?

Reply:
There are some calls to `apply_gate_BLAS` that specify other values for `state`, `pure` and `n` than the the corresponding class attributes in `measure_homodyne`. I haven't looked into it further though.